### PR TITLE
Add recursive spill for RowNumber

### PR DESCRIPTION
**Pull request by duanmeng**
At present, RowNumber only supports single-level spill partitions.
Once a spill takes place, it won't be spilled again. This means that
even if a restored spill partition uses an excessive amount of memory,
it still cannot be reclaimed. Therefore, it would be beneficial to implement
recursive spill support, as outlined in
https://facebookincubator.github.io/velox/develop/spilling.html.

1 Advance 'spillPartitionBits_' based on the partition bit of the currently
restoring spill partition to prepare for potential subsequent spills (recursive).

2 During the recursive spill input process, read the spilled input data from the
previous spill run via 'spillInputReader_', then spill it back into several sub-partitions.
Then restore one of the newly spilled partitions and reset 'spillInputReader' accordingly.

3 It's crucial to separate the recursive input spill process from the spill process itself,
as it can be time-consuming and should yield if it exceeds the CPU time limit.

In summary, the first spill operation writes all data to the disk. Subsequently,
RowNumber reads the data from the disk one partition at a time and clears
the data from the restored partition once it has been processed (#8413).
 In the case of a recursive spill, only the data from the currently restored partition
in memory is spilled to the disk (next level), and only one partition from the next
level is loaded back into memory.

The data layout may look like as follows.

||Memory|Disk|
| -- | -- | -- |
|No-Spill | [P0, P1]| []|
|First Spill | []| [P0, P1]|
|Restore | [P0] | [P1]|
|Second Spill | [] | [P0-0, P0-1, P1]|
|Restore| [P0-0] | [P0-1, P1]|
|Processed |[] |[P0-1, P1]|
|Restored |[P0-1] |[P1]|
|Processed | [] | [P1]|
|Restored | [P1] | []|

### CODE REVIEW

File: velox/exec/RowNumber.cpp [hunk 1]
```diff
@@ -444,6 +419,23 @@ void RowNumber::spill() {
     spillInput(input_, memory::spillMemoryPool());
     input_ = nullptr;
   }
+
+  // Recursive spill is triggered.
+  if (spillInputReader_ != nullptr) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:424]: This will take very long and we probably can't run this step during spill which run under memory arbitration.

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng thanks for the change and we might need to move the recursive input spill out of spill() call.

File: velox/exec/RowNumber.h [hunk 2]
```diff
@@ -78,6 +79,14 @@ class RowNumber : public Operator {
 
   FlatVector<int64_t>& getOrCreateRowNumberVector(vector_size_t size);
 
+  // Get partition start bit offset of the spill partitions. Return the value
+  // from spill config if it is the first spill. Advance (or right shift) the
+  // partition bit of the current restoring spill partition if the recursive
+  // spilling happens.
+  uint8_t startPartitionBit();
+
+  void recursiveSpillInput();
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.h:88]: Can you comment this function? Thanks!

File: velox/exec/RowNumber.cpp [hunk 3]
```diff
@@ -491,4 +470,44 @@ void RowNumber::spillInput(
...
+  const auto& spillConfig = spillConfig_.value();
+  return restoredSpillPartitionId_.has_value()
+      ? restoredSpillPartitionId_->partitionBitOffset() +
+          spillConfig.joinPartitionBits
+      : spillConfig.startPartitionBit;
+}
+
+void RowNumber::recursiveSpillInput() {
+  RowVectorPtr data;
+  while (spillInputReader_->nextBatch(data)) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:499]: Leave a TODO here to add yield handling? Thanks!

File: velox/exec/RowNumber.cpp [hunk 4]
```diff
@@ -491,4 +470,44 @@ void RowNumber::spillInput(
...
+uint8_t RowNumber::startPartitionBit() {
+  const auto& spillConfig = spillConfig_.value();
+  return restoredSpillPartitionId_.has_value()
+      ? restoredSpillPartitionId_->partitionBitOffset() +
+          spillConfig.joinPartitionBits
+      : spillConfig.startPartitionBit;
+}
+
+void RowNumber::recursiveSpillInput() {
+  RowVectorPtr data;
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:498]: s/data/input/

File: velox/exec/tests/RowNumberTest.cpp [hunk 5]
```diff
@@ -310,4 +326,168 @@ TEST_F(RowNumberTest, memoryUsage) {
   ASSERT_GE(peakBytesWithOutSpilling / peakBytesWithSpilling, 2);
 }
 
+DEBUG_ONLY_TEST_F(RowNumberTest, reclaimFromRowNumber) {
+  std::vector<RowVectorPtr> vectors = createVectors(8, rowType_, fuzzerOpts_);
+  createDuckDbTable(vectors);
+  std::unique_ptr<memory::MemoryManager> memoryManager = createMemoryManager();
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:334]: We can use shrinkPools from memory arbitrator for this test

File: velox/exec/tests/RowNumberTest.cpp [hunk 6]
```diff
@@ -310,4 +326,168 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+    auto task =
+        AssertQueryBuilder(duckDbQueryRunner_)
+            .spillDirectory(spillDirectory->path)
+            .config(core::QueryConfig::kSpillEnabled, true)
+            .config(core::QueryConfig::kRowNumberSpillEnabled, true)
+            .queryCtx(rowNumberQueryCtx)
+            .plan(PlanBuilder().values(vectors).rowNumber({"c0"}).planNode())
+            .assertResults(
+                "SELECT *, row_number() over (partition by c0) FROM tmp");
+    auto stats = task->taskStats().pipelineStats;
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:382]: use toPlanStats to verify spill stats? Thanks!

File: velox/exec/tests/RowNumberTest.cpp [hunk 7]
```diff
@@ -310,4 +326,168 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Task::requestPauseLocked",
+      std::function<void(Task*)>(([&](Task* /*unused*/) {
+        taskPauseWaitFlag = false;
+        taskPauseWait.notifyAll();
+      })));
+
+  std::atomic_int numOutputs{1};
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Driver::runInternal::getOutput",
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:446]: I think we just need a testValue of getOutput for this? And do we want a test to exceed the spill limit? Thanks!

File: velox/exec/RowNumber.cpp [hunk 8]
```diff
@@ -427,15 +398,23 @@ void RowNumber::setupInputSpiller() {
 
 void RowNumber::spill() {
   VELOX_CHECK(spillEnabled());
-  VELOX_CHECK_NULL(hashTableSpiller_);
-  VELOX_CHECK_NULL(inputSpiller_);
+  VELOX_CHECK(
+      (!hashTableSpiller_ && !inputSpiller_ && !spillInputReader_) ||
+      (hashTableSpiller_ && inputSpiller_ && spillInputReader_));
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/RowNumber.cpp:403]: Do you mean this check is redundant or just check hashTableSpiller？

File: velox/exec/RowNumber.cpp [hunk 9]
```diff
@@ -491,4 +470,44 @@ void RowNumber::spillInput(
...
+  return restoredSpillPartitionId_.has_value()
+      ? restoredSpillPartitionId_->partitionBitOffset() +
+          spillConfig.joinPartitionBits
+      : spillConfig.startPartitionBit;
+}
+
+void RowNumber::recursiveSpillInput() {
+  RowVectorPtr data;
+  while (spillInputReader_->nextBatch(data)) {
+    spillInput(data, memory::spillMemoryPool());
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:500]: Here we shall just use operator pool as it is not executed in spill() now

File: velox/exec/RowNumber.cpp [hunk 10]
```diff
@@ -491,4 +470,44 @@ void RowNumber::spillInput(
...
+  }
+
+  inputSpiller_->finishSpill(spillInputPartitionSet_);
+  spillInputReader_ = nullptr;
+
+  // recordSpillStats(hashTableSpiller_->stats());
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:506]: Remove this.

File: velox/exec/RowNumber.cpp [hunk 11]
```diff
@@ -427,15 +396,26 @@ void RowNumber::setupInputSpiller() {
...
+      (!hashTableSpiller_ && !inputSpiller_ && !spillInputReader_) ||
+      (hashTableSpiller_ && inputSpiller_ && spillInputReader_));
 
   setupHashTableSpiller();
   setupInputSpiller();
 
   hashTableSpiller_->spill();
   hashTableSpiller_->finishSpill(spillHashTablePartitionSet_);
 
+  // Records stats here when the recursive spill happens as the recursive spill
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:409]:
```
  // Records table spill stats if it is a recursive spill. The recursive spill can only happen during after noMoreInput stage.
  //
  // NOTE: we could only record spill stats after noMoreInput stage as the 'Operator::r...
  //
  // TODO: remove the no more input stage check in 'Operator::recordSpillStats' to ease implementation.
```

File: velox/exec/RowNumber.h [hunk 12]
```diff
@@ -78,6 +80,17 @@ class RowNumber : public Operator {
 
   FlatVector<int64_t>& getOrCreateRowNumberVector(vector_size_t size);
 
+  // Gets partition start bit offset for the next spill partition. Return the
+  // value from spill config if it is the first spill. Advance (or right shift)
+  // the partition bit of the current restoring spill partition in case of
+  // recursive spilling.
+  uint8_t startPartitionBit() const;
+
+  // Read previous spilled input data using the 'spillInputReader_' and spill
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.h:89]:
```
  // Used by recursive spill processing to read the spilled input data from the previous spill run through 'spillInputReader_' and then spill them back into a number of sub-partitions. After that, the function restores one of the newly spilled partition and reset 'spillInputReader_' accordingly.
```

File: velox/exec/tests/SpillTest.cpp [hunk 13]
```diff
@@ -740,6 +759,36 @@ TEST_P(SpillTest, nonExistSpillFileOnDeletion) {
   state_.reset();
 }
 
+TEST_P(SpillTest, removeEmptyPartitions) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/SpillTest.cpp:762]:
```
TEST(
```
This can be just tested once. Thanks!

File: velox/exec/tests/RowNumberTest.cpp [hunk 14]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
 }
 
+DEBUG_ONLY_TEST_F(RowNumberTest, spillOnlyDuringInputOrOutput) {
+  std::vector<RowVectorPtr> vectors = createVectors(8, rowType_, fuzzerOpts_);
+  createDuckDbTable(vectors);
+  std::unique_ptr<memory::MemoryManager> memoryManager = createMemoryManager();
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+
+  std::shared_ptr<core::QueryCtx> rowNumberQueryCtx;
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:336]:
```
nit: s/rowNumberQueryCtx/queryCtx/
```

File: velox/exec/tests/RowNumberTest.cpp [hunk 15]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+  rowNumberQueryCtx =
+      newQueryCtx(memoryManager, executor_, kMemoryCapacity * 2);
+
+  struct {
+    std::string spillStage;
+    std::string debugString() const {
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:342]: nit: leave an empty line in between?

s/spillStage/spillInjectionPoint/

File: velox/exec/tests/RowNumberTest.cpp [hunk 16]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+
+  std::shared_ptr<core::QueryCtx> rowNumberQueryCtx;
+  rowNumberQueryCtx =
+      newQueryCtx(memoryManager, executor_, kMemoryCapacity * 2);
+
+  struct {
+    int32_t numSpill;
+    int32_t maxSpillLevel;
+    std::string debugString() const {
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:408]: ditto

File: velox/exec/tests/RowNumberTest.cpp [hunk 17]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+  std::unique_ptr<memory::MemoryManager> memoryManager = createMemoryManager();
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+
+  std::shared_ptr<core::QueryCtx> rowNumberQueryCtx;
+  rowNumberQueryCtx =
+      newQueryCtx(memoryManager, executor_, kMemoryCapacity * 2);
+
+  struct {
+    int32_t numSpill;
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:406]:
```
s/numSpill/numSpills/
```

File: velox/exec/tests/RowNumberTest.cpp [hunk 18]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+        std::function<void(Operator*)>(([&](Operator* op) {
+          if (op->operatorType() != "RowNumber") {
+            return;
+          }
+
+          if (!op->testingNoMoreInput()) {
+            return;
+          }
+
+          if (numSpill >= testData.numSpill) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:444]:
```
if (numSpills++ >= testData.numSpill) {
  return;
}
```

**Review by xiaoxmeng — APPROVED** [review]
@duanmeng LGTM. Thanks!

File: velox/exec/tests/RowNumberTest.cpp [hunk 19]
```diff
@@ -310,4 +326,159 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+
+  std::shared_ptr<core::QueryCtx> queryCtx =
+      newQueryCtx(memoryManager, executor_, kMemoryCapacity * 2);
+
+  struct {
+    int32_t numSpills;
+    int32_t maxSpillLevel;
+
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:408]: Add int32_t expectedMaxSpilledLevel as test expectation? Thanks!

File: velox/exec/RowNumber.cpp [hunk 20]
```diff
@@ -378,8 +345,15 @@ void RowNumber::reclaim(
     return;
   }
 
-  if (hashTableSpiller_) {
-    // Already spilled.
+  // TODO: Add a spill level limit for RowNumber spill.
+  if (spillConfig_->exceedJoinSpillLevelLimit(startPartitionBit())) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:349]: We can add a flag in RowNumber for this and set it in setupHashTableSpiller?
startPartition_ bit which is set to the one in spill config initially and advance in setupHashTableSpiller?

Check if we have exceeded the max spill level in setupHashTableSpiller after we advance the start partition bit, and save in a flag exceededMaxSpillLevelLimit_.

Records this when we records the stats for hash table spiller? Thanks!

Comment (duanmeng) [reply velox/exec/RowNumber.cpp:349]: Keep the current implementation as is and simplify the recording of spill stats in the future per on our discussion.

File: velox/exec/tests/RowNumberTest.cpp [hunk 21]
```diff
@@ -310,4 +326,172 @@ TEST_F(RowNumberTest, memoryUsage) {
...
+        std::function<void(Operator*)>(([&](Operator* op) {
+          if (op->operatorType() != "RowNumber") {
+            return;
+          }
+
+          if (++numInputs != 5) {
+            return;
+          }
+
+          ++numSpills;
```
Comment (xiaoxmeng) [inline velox/exec/tests/RowNumberTest.cpp:430]: Remove this increment?

File: velox/exec/RowNumber.cpp [hunk 22]
```diff
@@ -491,4 +477,46 @@ void RowNumber::spillInput(
   }
 }
 
+void RowNumber::addInputInternal(const RowVectorPtr& input, bool fromSpill) {
+  const auto numInput = input->size();
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:481]: Shall we call ensureInputFits to prevent memory memory arbitration? Thanks!

Comment (duanmeng) [reply velox/exec/RowNumber.cpp:481]: It would be nice to add it. I will conduct a more proper way to add it in the follow-up.

File: velox/exec/tests/RowNumberTest.cpp (hunk 21 above)
Comment (duanmeng) [reply velox/exec/tests/RowNumberTest.cpp:435]: The first spill should be counted.

**Comment by duanmeng** [comment]
@xiaoxmeng Hi Xuan, I've made a small refactor based on your newly simplified `RowNumber` spill implementation. Could you please help to take another look? Thanks.

File: velox/exec/RowNumber.cpp (hunk 22 above)
Comment (xiaoxmeng) [reply velox/exec/RowNumber.cpp:488]: Leave a TODO here? Thanks!

File: velox/exec/RowNumber.cpp [hunk 23]
```diff
@@ -146,6 +114,17 @@ void RowNumber::restoreNextSpillPartition() {
   if (hashTableIt != spillHashTablePartitionSet_.end()) {
     spillHashTableReader_ = hashTableIt->second->createUnorderedReader(pool());
 
+    // Advance (or right shift) 'spillPartitionBits_' base on the partition bit
+    // of the current restoring spill partition in case of the next
+    // spill (recursive).
+    //
+    // Note: it may exceed the max spill level.
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:121]:
```
s/Note/NOTE/
```

File: velox/exec/RowNumber.cpp [hunk 24]
```diff
@@ -262,7 +241,11 @@ FlatVector<int64_t>& RowNumber::getOrCreateRowNumberVector(vector_size_t size) {
 
 RowVectorPtr RowNumber::getOutput() {
   if (input_ == nullptr) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:243]:
```
  if (input_ == nullptr) {
    if (spillInputReader_ != nullptr) {
      recursiveSpillInput();
    }
    if (input_ == nullptr)
      return nullptr;
    }
  }
```
Can we guarantee there is always data to restore here? Shall we add to check input_ after recursive spill input? thanks!

File: velox/exec/RowNumber.cpp [hunk 25]
```diff
@@ -440,16 +428,13 @@ void RowNumber::setupInputSpiller(
...
-
-  spillPartitionBits_ = HashBitRange(
-      spillConfig_->startPartitionBit,
-      spillConfig_->startPartitionBit + spillConfig_->joinPartitionBits);
 
   const auto spillPartitionSet = spillHashTable();
-
   setupInputSpiller(spillPartitionSet);
 
+  table_->clear();
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:435]: We already did this in spillHashTable()? Can we just check that after spillHashTable(), table_ is cleared?

File: velox/exec/RowNumber.cpp [hunk 26]
```diff
@@ -499,4 +471,60 @@ void RowNumber::spillInput(
...
+  recordSpillStats(inputSpiller_->stats());
+  spillInputReader_ = nullptr;
+
+  removeEmptyPartitions(spillInputPartitionSet_);
+  restoreNextSpillPartition();
+}
+
+void RowNumber::setSpillPartitionBits(
+    const SpillPartitionId* restoredPartitionId) {
+  if (restoredPartitionId == nullptr) {
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.cpp:511]:
```
const auto startPartitionBitOffset = restoredPartitionId == nullptr ? xxx : yyy;
spillPartitionBits_ = HashBitRange(...
```

File: velox/exec/RowNumber.h [hunk 27]
```diff
@@ -78,6 +80,18 @@ class RowNumber : public Operator {
 
   FlatVector<int64_t>& getOrCreateRowNumberVector(vector_size_t size);
 
+  // Used by recursive spill processing to read the spilled input data from the
+  // previous spill run through 'spillInputReader_' and then spill them back
+  // into a number of sub-partitions. After that, the function restores one of
+  // the newly spilled partitions and resets 'spillInputReader_' accordingly.
+  void recursiveSpillInput();
+
+  // Set 'spillPartitionBits_' in case of a spill (recursive).
```
Comment (xiaoxmeng) [inline velox/exec/RowNumber.h:89]:
```
// Set 'spillPartitionBits_' used for (recursive) spill
```

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng thanks for the update!

File: velox/exec/Spill.h [hunk 28]
```diff
@@ -456,6 +456,9 @@ class TestScopedSpillInjection {
 /// Test utility that returns true if triggered spill is evaluated to happen,
 /// false otherwise.
 bool testingTriggerSpill();
+
+/// Removes empty partitions from given spill partition set.
+void removeEmptyPartitions(SpillPartitionSet& partitionSet);
```
Comment (mbasmanova) [inline velox/exec/Spill.h:461]: This is a nice refactoring. Would it be possible to extract it into a separate PR to make this PR smaller and a bit easier to follow?

File: velox/exec/RowNumber.cpp [hunk 29]
```diff
@@ -379,8 +351,15 @@ void RowNumber::reclaim(
     return;
   }
 
-  if (inputSpiller_ != nullptr) {
-    // Already spilled.
+  // TODO: Add a spill level limit for RowNumber spill.
```
Comment (mbasmanova) [inline velox/exec/RowNumber.cpp:354]: Can we add it in this PR? It doesn't seem safe to not have a limit.

File: velox/exec/RowNumber.cpp [hunk 30]
```diff
@@ -379,8 +351,15 @@ void RowNumber::reclaim(
...
+  if (spillConfig_->exceedJoinSpillLevelLimit(spillPartitionBits_.begin())) {
```
Comment (mbasmanova) [inline velox/exec/RowNumber.cpp:355]: > exceedJoinSpillLevelLimit

Are we re-using join's limit? Let's add a limit for rownumber or rename join's config so it can be used in any operator.

File: velox/exec/RowNumber.cpp [hunk 31]
```diff
@@ -499,4 +471,60 @@ void RowNumber::spillInput(
   }
 }
 
+void RowNumber::addInputInternal(const RowVectorPtr& input, bool fromSpill) {
+  // TODO: Add ensureInputFits.
```
Comment (mbasmanova) [inline velox/exec/RowNumber.cpp:475]: Why is this a TODO?

File: velox/exec/RowNumber.cpp (hunk 30 above)
Comment (duanmeng) [reply velox/exec/RowNumber.cpp:355]: I've created a PR to config it, could you help to take a look? https://github.com/facebookincubator/velox/pull/8978

File: velox/exec/Spill.h (hunk 28 above)
Comment (duanmeng) [reply velox/exec/Spill.h:461]: Got it, Done. https://github.com/facebookincubator/velox/pull/9091

File: velox/exec/RowNumber.cpp (hunk 31 above)
Comment (duanmeng) [reply velox/exec/RowNumber.cpp:475]: Got it, I will add both `ensureInputFits` and `yield` in this PR.

File: velox/exec/RowNumber.cpp [hunk 32]
```diff
@@ -488,4 +473,74 @@ void RowNumber::spillInput(
...
+BlockingReason RowNumber::isBlocked(ContinueFuture* /* unused */) {
+  const auto reason =
+      yield_ ? BlockingReason::kYield : BlockingReason::kNotBlocked;
+  yield_ = false;
+  return reason;
+}
+
+void RowNumber::addInputInternal(const RowVectorPtr& input, bool fromSpill) {
+  ensureInputFits(input);
+  if (!fromSpill && inputSpiller_ != nullptr) {
```
Comment (bikramSingh91) [inline velox/exec/RowNumber.cpp:485]: do we expect to encounter this condition in any case? if not, should this be a VELOX_CHECK instead of a no-op return ?

File: velox/exec/RowNumber.cpp [hunk 33]
```diff
@@ -488,4 +473,74 @@ void RowNumber::spillInput(
...
+    return;
+  }
+
+  if (input == nullptr) {
```
Comment (bikramSingh91) [inline velox/exec/RowNumber.cpp:489]: ditto as comment on L485

File: velox/exec/RowNumber.cpp [hunk 34]
```diff
@@ -250,7 +225,19 @@ FlatVector<int64_t>& RowNumber::getOrCreateRowNumberVector(vector_size_t size) {
 
 RowVectorPtr RowNumber::getOutput() {
   if (input_ == nullptr) {
-    return nullptr;
+    if (spillInputReader_ == nullptr) {
+      return nullptr;
+    }
+
+    if (yield_) {
+      VELOX_CHECK_NULL(input_);
```
Comment (bikramSingh91) [inline velox/exec/RowNumber.cpp:233]: nit: do we need this check? since we only reach here if condition on L227 is true where input_ == nullptr

**Review by bikramSingh91 — COMMENTED** [review]
Looks good, just a few nits. Thank you for adding recursive spilling to this Operators.

Can you please also update Operators.rst(L919) to mention that RowNumber Operator now supports spilling.

**Review by duanmeng — COMMENTED** [review]
@bikramSingh91 Thanks for your review, I've resolved all the comments and updated the doc. Could you help take another look? Thanks a lot.

File: velox/exec/RowNumber.cpp [hunk 35]
```diff
@@ -97,8 +101,14 @@ void RowNumber::addInput(RowVectorPtr input) {
 }
 
 void RowNumber::addSpillInput() {
+  ensureInputFits(input_);
```
Comment (bikramSingh91) [inline velox/exec/RowNumber.cpp:104]: this will be called twice when invoked via addInput(), will this result in the reservation being increased twice?

Comment (duanmeng) [reply velox/exec/RowNumber.cpp:104]: No, the `addInput` is called before `noMoreInput`, while `addInputSpill` is called after `noMoreInput`. 
The `addInputSpill` is specifically called right after the spillInputReader_ reads a batch in either the `restoreNextSpillPartition` or `getOutput`. The spillInputReader_ is null before `noMoreInput` is called, and it first becomes non-null at the end of `restoreNextSpillPartition` when the `noMoreInput` is called.

**Comment by duanmeng** [comment]
@bikramSingh91 Hi Bikram, any further comments? Thanks.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, all the prerequisite PRs have been merged, could you please help take another look? Thanks.

**Comment by mbasmanova** [comment]
@duanmeng Thank you for adding "The data layout may look like as follows." section to PR description. It is very helpful.

Spilling is rather complicated and we are seeing many bugs in it. I wonder how much testing have been done for this change and what's the confidence of it being bug free. I don't believe we have fuzzer coverage for RowNumber operator. Would that be necessary to ensure there are no (or very few) bugs?

Also, I wonder what's the motivation for this change. Is there a real-world scenario that benefits from this change and justifies added complexity and maintenance costs. Would you describe that scenario? In this scenario, what's the impact on latency and CPU usage of spilling? How much CPU and wall time a query would use without spilling and how does that compare to spilling?

File: velox/common/base/SpillConfig.h [hunk 36]
```diff
@@ -69,8 +69,8 @@ struct SpillConfig {
   /// the next level of recursive spilling.
   int32_t spillLevel(uint8_t startBitOffset) const;
 
-  /// Checks if the given 'startBitOffset' and 'numPartitionBits' has exceeded
-  /// the max hash join spill limit.
+  /// Checks if the given 'startBitOffset' has exceeded the max hash join spill
```
Comment (mbasmanova) [inline velox/common/base/SpillConfig.h:72]: Why is this comment referring to hash join? Is this a typo?

File: velox/exec/RowNumber.cpp [hunk 37]
```diff
@@ -489,4 +508,46 @@ void RowNumber::spillInput(
   }
 }
 
+BlockingReason RowNumber::isBlocked(ContinueFuture* /* unused */) {
+  const auto reason =
+      yield_ ? BlockingReason::kYield : BlockingReason::kNotBlocked;
```
Comment (mbasmanova) [inline velox/exec/RowNumber.cpp:513]: This logic is a bit strange. isBlocked is supposed to set the 'future' if operator is blocked, but this code doesn't do that.

File: velox/exec/RowNumber.cpp [hunk 38]
```diff
@@ -489,4 +508,46 @@ void RowNumber::spillInput(
...
+  yield_ = false;
+  return reason;
+}
+
+void RowNumber::recursiveSpillInput() {
+  RowVectorPtr input;
+  while (spillInputReader_->nextBatch(input)) {
+    spillInput(input, pool());
+
+    if (operatorCtx_->driver()->shouldYield()) {
```
Comment (mbasmanova) [inline velox/exec/RowNumber.cpp:523]: Is there test coverage for this code path?

File: velox/common/base/SpillConfig.h (hunk 36 above)
Comment (duanmeng) [reply velox/common/base/SpillConfig.h]: Fixed.

File: velox/exec/RowNumber.cpp (hunk 37 above)
Comment (duanmeng) [reply velox/exec/RowNumber.cpp:513]: We need to return BlockingReason::kYield in `isBlocked` to make the driver framework aware to off the thread and yield the CPU.
https://github.com/facebookincubator/velox/blob/main/velox/exec/Driver.cpp#L545
https://github.com/facebookincubator/velox/blob/main/velox/exec/Driver.cpp#L776

Comment (duanmeng) [reply velox/exec/RowNumber.cpp]: Fixed. Yes, it is kind of hacky. Turns out we could leverage the **yield** in the driver framework itself.
https://github.com/facebookincubator/velox/blob/main/velox/exec/Driver.cpp#L526

Comment (mbasmanova) [reply velox/exec/RowNumber.cpp]: @duanmeng Thanks. Should we add a check that if isBlocked returns something other than kNotBlocked, then future is set? (in a separate PR)

Comment (duanmeng) [reply velox/exec/RowNumber.cpp]: @mbasmanova I am not sure whether we need this check as there is no blocking future in `RowNumber` set the 'future' of the `isBlocked` API.

Comment (mbasmanova) [reply velox/exec/RowNumber.cpp]: I was thinking that the caller of the Operator::isBlocked API may need to add a check that if the API returned "blocked", then the future is set.

Comment (duanmeng) [reply velox/exec/RowNumber.cpp]: Do you mean we need to add a check in the Driver::runInternal to ensure that if the `op->isBlocked` does not return `kNotBlocked` then the `future` is not empty? So that this can regulate developers to ensure that the future is correctly set when overriding the `isBlocked` API.

Comment (mbasmanova) [reply velox/exec/RowNumber.cpp]: This is what I'm thinking about.

**Review by tanjialiang — COMMENTED** [review]
Do we want to add a flat to enable/disable the spill of this operator, like join and aggregation?
